### ISSUE TITLE
Queries using a dashboard variable as datasource do not update on variable change

### ISSUE THREAD

**Issue by PaulRitter** [grafana/grafana#56987]
I have a dashboard variable defined as follows:
```json
{
        "current": {
          "selected": true,
          "text": "Content Benchmarks PostgreSQL ",
          "value": "Content Benchmarks PostgreSQL "
        },
        "hide": 0,
        "includeAll": false,
        "label": "Database",
        "multi": false,
        "name": "database",
        "options": [],
        "query": "postgres",
        "queryValue": "",
        "refresh": 1,
        "regex": ".*Benchmarks.*",
        "skipUrlSync": false,
        "type": "datasource"
      }
```
And am using it in my query of my panel:
```json
{
      "datasource": {
        "type": "postgres",
        "uid": "${database}"
      },
...
```

**What happened**:
I change the datasource parameter value, but the datasource of the query in my panel did not change.

**What you expected to happen**:
For the datasource of the panel query to change.

**How to reproduce it (as minimally and precisely as possible)**:
See above

**Anything else we need to know?**:
Editing the panel and reselecting the ${database} datasource value updates the query to the correct value.

**Environment**:
- Grafana version: v8.5.2 (90701be19)
- Data source type & version: PostgreSQL 13.6
- OS Grafana is installed on: Linux hostname 4.18.0-240.1.1.el8_3.x86_64 \#\1 SMP Thu Nov 19 17:20:08 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
- User OS & Browser: Firefox 105.0.3
- Grafana plugins: /
- Others: /

**Comment by ashharrison90** [comment]
hi @PaulRitter 👋 

can't quite manage to reproduce this on v8.5.2 locally:

https://user-images.githubusercontent.com/20999846/196162836-a80bd547-f2fb-4d50-b922-959a241abc12.mp4


is there any other information you can provide that might help us debug this issue? 🤔

**Comment by PaulRitter** [comment]
Interesting. Here is the full dashboard json, hope it helps:
```json
{
  "annotations": {
    "list": [
      {
        "builtIn": 1,
        "datasource": {
          "type": "grafana",
          "uid": "-- Grafana --"
        },
        "enable": true,
        "hide": true,
        "iconColor": "rgba(0, 211, 255, 1)",
        "name": "Annotations & Alerts",
        "target": {
          "limit": 100,
          "matchAny": false,
          "tags": [],
          "type": "dashboard"
        },
        "type": "dashboard"
      }
    ]
  },
  "editable": true,
  "fiscalYearStartMonth": 0,
  "graphTooltip": 0,
  "id": 20,
  "iteration": 1666006179445,
  "links": [],
  "liveNow": false,
  "panels": [
    {
      "datasource": {
        "type": "postgres",
        "uid": "${database}"
      },
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "palette-classic"
          },
          "custom": {
            "axisGridShow": true,
            "axisLabel": "Completion Time",
            "axisPlacement": "auto",
            "axisSoftMin": 0,
            "barAlignment": 0,
            "drawStyle": "line",
            "fillOpacity": 0,
            "gradientMode": "none",
            "hideFrom": {
              "legend": false,
              "tooltip": false,
              "viz": false
            },
            "lineInterpolation": "linear",
            "lineStyle": {
              "fill": "solid"
            },
            "lineWidth": 2,
            "pointSize": 10,
            "scaleDistribution": {
              "type": "linear"
            },
            "showPoints": "always",
            "spanNulls": true,
            "stacking": {
              "group": "A",
              "mode": "none"
            },
            "thresholdsStyle": {
              "mode": "off"
            }
          },
          "mappings": [],
          "thresholds": {
            "mode": "percentage",
            "steps": [
              {
                "color": "green",
                "value": null
              }
            ]
          },
          "unit": "ns"
        },
        "overrides": []
      },
      "gridPos": {
        "h": 7,
        "w": 12,
        "x": 0,
        "y": 0
      },
      "id": 6,
      "options": {
        "legend": {
          "calcs": [],
          "displayMode": "list",
          "placement": "bottom"
        },
        "tooltip": {
          "mode": "multi",
          "sort": "asc"
        }
      },
      "pluginVersion": "8.5.2",
      "targets": [
        {
          "datasource": {
            "type": "postgres",
            "uid": "VfnrlVIVk"
          },
          "format": "table",
          "group": [],
          "metricColumn": "\"Name\"",
          "rawQuery": true,
          "rawSql": "SELECT\n  \"RunDate\" AS \"time\",\n  \"ParameterMapping\" AS metric,\n  \"Statistics\"\nFROM \"BenchmarkRuns\"\nWHERE\n  $__timeFilter(\"RunDate\") AND\n  \"Name\" = '$benchmark_name' AND\n  (\"ParameterMapping\" IN ($parameter_mapping) OR \"ParameterMapping\" IS NULL)\nORDER BY 1,2",
          "refId": "A",
          "select": [
            [
              {
                "params": [
                  "\"Reports\""
                ],
                "type": "column"
              }
            ]
          ],
          "table": "\"BenchmarkRuns\"",
          "timeColumn": "\"RunDate\"",
          "timeColumnType": "timestamptz",
          "where": [
            {
              "name": "$__timeFilter",
              "params": [],
              "type": "macro"
            },
            {
              "datatype": "text",
              "name": "",
              "params": [
                "\"Name\"",
                "=",
                "'$benchmark_name'"
              ],
              "type": "expression"
            }
          ]
        }
      ],
      "title": "Mean",
      "transformations": [
        {
          "id": "extractFields",
          "options": {
            "format": "json",
            "replace": false,
            "source": "Statistics"
          }
        },
        {
          "id": "organize",
          "options": {
            "excludeByName": {
              "AllOutliers": true,
              "ConfidenceInterval": true,
              "InterquartileRange": true,
              "Kurtosis": true,
              "LowerFence": true,
              "LowerOutliers": true,
              "Max": true,
              "Mean": false,
              "Median": true,
              "Min": true,
              "N": true,
              "OriginalValues": true,
              "Percentiles": true,
              "Q1": true,
              "Q3": true,
              "Skewness": true,
              "StandardDeviation": true,
              "StandardError": true,
              "Statistics": true,
              "UpperFence": true,
              "UpperOutliers": true,
              "Variance": true,
              "metric": false
            },
            "indexByName": {},
            "renameByName": {
              "Mean": ""
            }
          }
        },
        {
          "id": "prepareTimeSeries",
          "options": {
            "format": "many"
          }
        }
      ],
      "type": "timeseries"
    },
    {
      "datasource": {
        "type": "postgres",
        "uid": "${database}"
      },
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "palette-classic"
          },
          "custom": {
            "axisGridShow": true,
            "axisLabel": "Completion Time",
            "axisPlacement": "auto",
            "axisSoftMin": 0,
            "barAlignment": 0,
            "drawStyle": "line",
            "fillOpacity": 0,
            "gradientMode": "none",
            "hideFrom": {
              "legend": false,
              "tooltip": false,
              "viz": false
            },
            "lineInterpolation": "linear",
            "lineStyle": {
              "fill": "solid"
            },
            "lineWidth": 2,
            "pointSize": 10,
            "scaleDistribution": {
              "type": "linear"
            },
            "showPoints": "always",
            "spanNulls": true,
            "stacking": {
              "group": "A",
              "mode": "none"
            },
            "thresholdsStyle": {
              "mode": "off"
            }
          },
          "mappings": [],
          "thresholds": {
            "mode": "percentage",
            "steps": [
              {
                "color": "green",
                "value": null
              }
            ]
          },
          "unit": "ns"
        },
        "overrides": []
      },
      "gridPos": {
        "h": 7,
        "w": 12,
        "x": 12,
        "y": 0
      },
      "id": 5,
      "options": {
        "legend": {
          "calcs": [],
          "displayMode": "list",
          "placement": "bottom"
        },
        "tooltip": {
          "mode": "multi",
          "sort": "asc"
        }
      },
      "pluginVersion": "8.5.2",
      "targets": [
        {
          "datasource": {
            "type": "postgres",
            "uid": "7l9pCQ_nz"
          },
          "format": "table",
          "group": [],
          "metricColumn": "\"Name\"",
          "rawQuery": true,
          "rawSql": "SELECT\n  \"RunDate\" AS \"time\",\n  \"ParameterMapping\" AS metric,\n  \"Statistics\"\nFROM \"BenchmarkRuns\"\nWHERE\n  $__timeFilter(\"RunDate\") AND\n  \"Name\" = '$benchmark_name' AND\n  (\"ParameterMapping\" IN ($parameter_mapping) OR \"ParameterMapping\" IS NULL)\nORDER BY 1,2",
          "refId": "A",
          "select": [
            [
              {
                "params": [
                  "\"Reports\""
                ],
                "type": "column"
              }
            ]
          ],
          "table": "\"BenchmarkRuns\"",
          "timeColumn": "\"RunDate\"",
          "timeColumnType": "timestamptz",
          "where": [
            {
              "name": "$__timeFilter",
              "params": [],
              "type": "macro"
            },
            {
              "datatype": "text",
              "name": "",
              "params": [
                "\"Name\"",
                "=",
                "'$benchmark_name'"
              ],
              "type": "expression"
            }
          ]
        }
      ],
      "title": "Standard Deviation",
      "transformations": [
        {
          "id": "extractFields",
          "options": {
            "format": "json",
            "replace": false,
            "source": "Statistics"
          }
        },
        {
          "id": "organize",
          "options": {
            "excludeByName": {
              "AllOutliers": true,
              "ConfidenceInterval": true,
              "InterquartileRange": true,
              "Kurtosis": true,
              "LowerFence": true,
              "LowerOutliers": true,
              "Max": true,
              "Mean": true,
              "Median": true,
              "Min": true,
              "N": true,
              "OriginalValues": true,
              "Percentiles": true,
              "Q1": true,
              "Q3": true,
              "Skewness": true,
              "StandardDeviation": false,
              "StandardError": true,
              "Statistics": true,
              "UpperFence": true,
              "UpperOutliers": true,
              "Variance": true,
              "metric": false
            },
            "indexByName": {},
            "renameByName": {}
          }
        },
        {
          "id": "prepareTimeSeries",
          "options": {
            "format": "many"
          }
        }
      ],
      "type": "timeseries"
    },
    {
      "datasource": {
        "type": "postgres",
        "uid": "${database}"
      },
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "palette-classic"
          },
          "custom": {
            "axisGridShow": true,
            "axisLabel": "Completion Time",
            "axisPlacement": "auto",
            "axisSoftMin": 0,
            "barAlignment": 0,
            "drawStyle": "line",
            "fillOpacity": 0,
            "gradientMode": "none",
            "hideFrom": {
              "legend": false,
              "tooltip": false,
              "viz": false
            },
            "lineInterpolation": "linear",
            "lineStyle": {
              "fill": "solid"
            },
            "lineWidth": 2,
            "pointSize": 10,
            "scaleDistribution": {
              "type": "linear"
            },
            "showPoints": "always",
            "spanNulls": true,
            "stacking": {
              "group": "A",
              "mode": "none"
            },
            "thresholdsStyle": {
              "mode": "off"
            }
          },
          "mappings": [],
          "thresholds": {
            "mode": "percentage",
            "steps": [
              {
                "color": "green",
                "value": null
              }
            ]
          },
          "unit": "ns"
        },
        "overrides": []
      },
      "gridPos": {
        "h": 7,
        "w": 12,
        "x": 0,
        "y": 7
      },
      "id": 4,
      "options": {
        "legend": {
          "calcs": [],
          "displayMode": "list",
          "placement": "bottom"
        },
        "tooltip": {
          "mode": "multi",
          "sort": "asc"
        }
      },
      "pluginVersion": "8.5.2",
      "targets": [
        {
          "datasource": {
            "type": "postgres",
            "uid": "7l9pCQ_nz"
          },
          "format": "table",
          "group": [],
          "metricColumn": "\"Name\"",
          "rawQuery": true,
          "rawSql": "SELECT\n  \"RunDate\" AS \"time\",\n  \"ParameterMapping\" AS metric,\n  \"Statistics\"\nFROM \"BenchmarkRuns\"\nWHERE\n  $__timeFilter(\"RunDate\") AND\n  \"Name\" = '$benchmark_name' AND\n  (\"ParameterMapping\" IN ($parameter_mapping) OR \"ParameterMapping\" IS NULL)\nORDER BY 1,2",
          "refId": "A",
          "select": [
            [
              {
                "params": [
                  "\"Reports\""
                ],
                "type": "column"
              }
            ]
          ],
          "table": "\"BenchmarkRuns\"",
          "timeColumn": "\"RunDate\"",
          "timeColumnType": "timestamptz",
          "where": [
            {
              "name": "$__timeFilter",
              "params": [],
              "type": "macro"
            },
            {
              "datatype": "text",
              "name": "",
              "params": [
                "\"Name\"",
                "=",
                "'$benchmark_name'"
              ],
              "type": "expression"
            }
          ]
        }
      ],
      "title": "Median",
      "transformations": [
        {
          "id": "extractFields",
          "options": {
            "format": "json",
            "replace": false,
            "source": "Statistics"
          }
        },
        {
          "id": "organize",
          "options": {
            "excludeByName": {
              "AllOutliers": true,
              "ConfidenceInterval": true,
              "InterquartileRange": true,
              "Kurtosis": true,
              "LowerFence": true,
              "LowerOutliers": true,
              "Max": true,
              "Mean": true,
              "Min": true,
              "N": true,
              "OriginalValues": true,
              "Percentiles": true,
              "Q1": true,
              "Q3": true,
              "Skewness": true,
              "StandardDeviation": true,
              "StandardError": true,
              "Statistics": true,
              "UpperFence": true,
              "UpperOutliers": true,
              "Variance": true,
              "metric": false
            },
            "indexByName": {},
            "renameByName": {}
          }
        },
        {
          "id": "prepareTimeSeries",
          "options": {
            "format": "many"
          }
        }
      ],
      "type": "timeseries"
    },
    {
      "datasource": {
        "type": "postgres",
        "uid": "${database}"
      },
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "palette-classic"
          },
          "custom": {
            "axisGridShow": true,
            "axisLabel": "Completion Time",
            "axisPlacement": "auto",
            "axisSoftMin": 0,
            "barAlignment": 0,
            "drawStyle": "line",
            "fillOpacity": 0,
            "gradientMode": "none",
            "hideFrom": {
              "legend": false,
              "tooltip": false,
              "viz": false
            },
            "lineInterpolation": "linear",
            "lineStyle": {
              "fill": "solid"
            },
            "lineWidth": 2,
            "pointSize": 10,
            "scaleDistribution": {
              "type": "linear"
            },
            "showPoints": "always",
            "spanNulls": true,
            "stacking": {
              "group": "A",
              "mode": "none"
            },
            "thresholdsStyle": {
              "mode": "off"
            }
          },
          "mappings": [],
          "thresholds": {
            "mode": "percentage",
            "steps": [
              {
                "color": "green",
                "value": null
              }
            ]
          },
          "unit": "ns"
        },
        "overrides": []
      },
      "gridPos": {
        "h": 7,
        "w": 12,
        "x": 12,
        "y": 7
      },
      "id": 7,
      "options": {
        "legend": {
          "calcs": [],
          "displayMode": "list",
          "placement": "bottom"
        },
        "tooltip": {
          "mode": "multi",
          "sort": "asc"
        }
      },
      "pluginVersion": "8.5.2",
      "targets": [
        {
          "datasource": {
            "type": "postgres",
            "uid": "7l9pCQ_nz"
          },
          "format": "table",
          "group": [],
          "metricColumn": "\"Name\"",
          "rawQuery": true,
          "rawSql": "SELECT\n  \"RunDate\" AS \"time\",\n  \"ParameterMapping\" AS metric,\n  \"Statistics\"\nFROM \"BenchmarkRuns\"\nWHERE\n  $__timeFilter(\"RunDate\") AND\n  \"Name\" = '$benchmark_name' AND\n  (\"ParameterMapping\" IN ($parameter_mapping) OR \"ParameterMapping\" IS NULL)\nORDER BY 1,2",
          "refId": "A",
          "select": [
            [
              {
                "params": [
                  "\"Reports\""
                ],
                "type": "column"
              }
            ]
          ],
          "table": "\"BenchmarkRuns\"",
          "timeColumn": "\"RunDate\"",
          "timeColumnType": "timestamptz",
          "where": [
            {
              "name": "$__timeFilter",
              "params": [],
              "type": "macro"
            },
            {
              "datatype": "text",
              "name": "",
              "params": [
                "\"Name\"",
                "=",
                "'$benchmark_name'"
              ],
              "type": "expression"
            }
          ]
        }
      ],
      "title": "Standard Error",
      "transformations": [
        {
          "id": "extractFields",
          "options": {
            "format": "json",
            "replace": false,
            "source": "Statistics"
          }
        },
        {
          "id": "organize",
          "options": {
            "excludeByName": {
              "AllOutliers": true,
              "ConfidenceInterval": true,
              "InterquartileRange": true,
              "Kurtosis": true,
              "LowerFence": true,
              "LowerOutliers": true,
              "Max": true,
              "Mean": true,
              "Median": true,
              "Min": true,
              "N": true,
              "OriginalValues": true,
              "Percentiles": true,
              "Q1": true,
              "Q3": true,
              "Skewness": true,
              "StandardDeviation": true,
              "StandardError": false,
              "Statistics": true,
              "UpperFence": true,
              "UpperOutliers": true,
              "Variance": true,
              "metric": false
            },
            "indexByName": {},
            "renameByName": {}
          }
        },
        {
          "id": "prepareTimeSeries",
          "options": {
            "format": "many"
          }
        }
      ],
      "type": "timeseries"
    }
  ],
  "refresh": "",
  "schemaVersion": 36,
  "style": "dark",
  "tags": [],
  "templating": {
    "list": [
      {
        "current": {
          "selected": false,
          "text": "Benchmarks PostgreSQL",
          "value": "Benchmarks PostgreSQL"
        },
        "hide": 0,
        "includeAll": false,
        "label": "Database",
        "multi": false,
        "name": "database",
        "options": [],
        "query": "postgres",
        "queryValue": "",
        "refresh": 1,
        "regex": ".*Benchmarks.*",
        "skipUrlSync": false,
        "type": "datasource"
      },
      {
        "current": {
          "selected": false,
          "text": "SerializationArrayBenchmark.ReadTenSealedStringDataDefs",
          "value": "SerializationArrayBenchmark.ReadTenSealedStringDataDefs"
        },
        "datasource": {
          "type": "postgres",
          "uid": "${database}"
        },
        "definition": "SELECT DISTINCT \"Name\" FROM \"BenchmarkRuns\" WHERE $__timeFilter(\"RunDate\")",
        "hide": 0,
        "includeAll": false,
        "label": "Benchmark Name",
        "multi": false,
        "name": "benchmark_name",
        "options": [],
        "query": "SELECT DISTINCT \"Name\" FROM \"BenchmarkRuns\" WHERE $__timeFilter(\"RunDate\")",
        "refresh": 2,
        "regex": "",
        "skipUrlSync": false,
        "sort": 0,
        "type": "query"
      },
      {
        "current": {
          "selected": false,
          "text": "All",
          "value": "$__all"
        },
        "datasource": {
          "type": "postgres",
          "uid": "${database}"
        },
        "definition": "SELECT DISTINCT \"ParameterMapping\" FROM \"BenchmarkRuns\" WHERE $__timeFilter(\"RunDate\") AND \"Name\" = '$benchmark_name'",
        "hide": 0,
        "includeAll": true,
        "label": "Parameter Mapping",
        "multi": true,
        "name": "parameter_mapping",
        "options": [],
        "query": "SELECT DISTINCT \"ParameterMapping\" FROM \"BenchmarkRuns\" WHERE $__timeFilter(\"RunDate\") AND \"Name\" = '$benchmark_name'",
        "refresh": 2,
        "regex": "",
        "skipUrlSync": false,
        "sort": 0,
        "type": "query"
      }
    ]
  },
  "time": {
    "from": "now-30d",
    "to": "now"
  },
  "timepicker": {},
  "timezone": "",
  "title": "Benchmarks",
  "uid": "a88u7_l7k",
  "version": 25,
  "weekStart": ""
}
```

**Comment by PaulRitter** [comment]
I was able to resolve this issue, however it is still a bug:
The dashboard was originally used without the datasource parameter. Apparently, while being converted, at some point down the way, grafana wrote down these uids (7l9pCQ_nz, VfnrlVIVk) instead of the parameter uid into the uid fields of the specific targets (Note: even though the ${database} parameter was selected!) which caused a "deadlock" of grafana not using the database-variable even when selecting it while editing the panel. Apparently, everytime you would select the parameter & save, it would grab the uid of the currently selected table and save that, instead of saving the variable reference uid.

I fixed this by manually replacing the fields with the variable reference:
7l9pCQ_nz -> ${database}
VfnrlVIVk -> ${database}
